src/PowerShellEditorServices/Workspace/ScriptFile.cs

```diff
@@ -347,51 +373,65 @@ public void ApplyChange(FileChange fileChange)
-                    this.FileLines.RemoveAt(fileChange.Line - 1);
-                }
+                this.ValidatePosition(fileChange.Line, fileChange.Offset, isInsertion: true);
+                this.ValidatePosition(fileChange.EndLine, fileChange.EndOffset, isInsertion: true);
 
-                // Build and insert the new lines
-                int currentLineNumber = fileChange.Line;
-                for (int changeIndex = 0; changeIndex < changeLines.Length; changeIndex++)
+                // VSCode sometimes likes to give the change start line as (FileLines.Count + 1).
+                // This used to crash EditorServices, but we now treat it as an append.
+                // See https://github.com/PowerShell/vscode-powershell/issues/1283
+                if (fileChange.Line == this.FileLines.Count + 1)
                 {
-                    // Since we split the lines above using \n, make sure to
-                    // trim the ending \r's off as well.
-                    string finalLine = changeLines[changeIndex].TrimEnd('\r');
-
-                    // Should we add first or last line fragments?
-                    if (changeIndex == 0)
+                    foreach (string addedLine in changeLines)
                     {
-                        // Append the first line fragment
-                        finalLine = firstLineFragment + finalLine;
+                        string finalLine = addedLine.TrimEnd('\r');
+                        this.FileLines.Add(finalLine);
                     }
-                    if (changeIndex == changeLines.Length - 1)
+                }
+                // Otherwise, the change needs to go between existing content
+                else
+                {
+                    // Get the first fragment of the first line
+                    string firstLineFragment =
+                    this.FileLines[fileChange.Line - 1]
+                        .Substring(0, fileChange.Offset - 1);
+
+                    // Get the last fragment of the last line
+                    string endLine = this.FileLines[fileChange.EndLine - 1];
+                    string lastLineFragment =
+                    endLine.Substring(
+                        fileChange.EndOffset - 1,
+                        (this.FileLines[fileChange.EndLine - 1].Length - fileChange.EndOffset) + 1);
+
+                    // Remove the old lines
+                    for (int i = 0; i <= fileChange.EndLine - fileChange.Line; i++)
                     {
-                        // Append the last line fragment
-                        finalLine = finalLine + lastLineFragment;
+                        this.FileLines.RemoveAt(fileChange.Line - 1);
                     }
 
-                    this.FileLines.Insert(currentLineNumber - 1, finalLine);
-                    currentLineNumber++;
-                }
+                    // Build and insert the new lines
+                    int currentLineNumber = fileChange.Line;
```

Comment:
maybe consider doing the -1 here instead of line 431. I think it's a little clearer.